src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs

```diff
@@ -202,57 +200,76 @@ await client.TryInvokeAsync<IRemoteAssetSynchronizationService>(
-        // without any high cost and create new snapshot from it.
-        //
-        // otherwise, it will do the normal behavior of getting full text from VS side. this optimization saves
-        // times we need to do full text synchronization for typing scenario.
-        using var _ = ArrayBuilder<(DocumentId id, Checksum textChecksum, ImmutableArray<TextChange> changes, Checksum newTextChecksum)>.GetInstance(out var builder);
-
-        foreach (var (oldDocument, newDocument) in values)
+        // Attempt to inform the remote asset synchronization service as quickly as possible
+        // about the text changes between oldDocument and newDocument. By doing this, we can
+        // reduce the likelihood of the remote side encountering an unknown checksum and
+        // requiring a synchronization of the full document contents.
+        // This method uses JTF.RunAsync to create a fire-and-forget task. JTF.RunAsync will
+        // execute DispatchSynchronizeTextChangesHelperAsync synchronously until no longer
+        // possible. The hopefully common occurrence is that it is able to run synchronously
+        // all the way until it fires off the RPC call notifying the remote side of the changes.
+        _ = _threadingContext.JoinableTaskFactory.RunAsync(async () =>
         {
-            if (!oldDocument.TryGetText(out var oldText) ||
-                !newDocument.TryGetText(out var newText))
-            {
-                // we only support case where text already exist
-                continue;
-            }
-
-            // Avoid allocating text before seeing if we can bail out.
-            var changeRanges = newText.GetChangeRanges(oldText).AsImmutable();
-            if (changeRanges.Length == 0)
-                continue;
-
-            // no benefit here. pulling from remote host is more efficient
-            if (changeRanges is [{ Span.Length: var singleChangeLength }] && singleChangeLength == oldText.Length)
-                continue;
+            var wasSynchronized = await DispatchSynchronizeTextChangesHelperAsync().ConfigureAwait(false);
 
-            var state = await oldDocument.State.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-            var newState = await newDocument.State.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-
-            var textChanges = newText.GetTextChanges(oldText).AsImmutable();
-            builder.Add((oldDocument.Id, state.Text, textChanges, newState.Text));
-        }
+            var metricName = wasSynchronized ? "SucceededCount" : "FailedCount";
+            TelemetryLogging.LogAggregatedCounter(FunctionId.ChecksumUpdater_SynchronizeTextChangesStatus, KeyValueLogMessage.Create(m =>
+            {
+                m[TelemetryLogging.KeyName] = nameof(SolutionChecksumUpdater) + "." + metricName;
+                m[TelemetryLogging.KeyValue] = 1L;
+                m[TelemetryLogging.KeyMetricName] = metricName;
```

Comment:
why do we need this here, if it is encoded in KeyName above? is one of these redundant?

Reply:
This ends up down in AbstractAggregatingLog.Log. That code requires KeyName/KeyValue to be specified, and if KeyMetricName isn't specified, falls back to KeyName to be the metricName.

I can look into whether there are callers that depend on that behavior, and if not, remove that parameter. However, I'd prefer not to as even if that could be removed, it would add an allocation to strip out the metricname from the keyname.